### PR TITLE
Medical GUI - Peek Medical Info on Hit

### PR DESCRIPTION
**Pull request by amsteadrayle**
**When merged this pull request will:**
- Temporarily show the player's patient information display when hit

### IMPORTANT

- [ ] If the contribution affects [the documentation](https://github.com/acemod/ACE3/tree/master/docs), please include your changes in this pull request so the documentation will appear on the [website](https://ace3.acemod.org/).
- [x] [Development Guidelines](https://ace3.acemod.org/wiki/development/) are read, understood and applied.
- [x] Title of this PR uses our standard template `Component - Add|Fix|Improve|Change|Make|Remove {changes}`.

### CODE REVIEW

**Comment by amsteadrayle** [comment]
Like with my other PR, the goal with this is to provide the player with immediate feedback about their medical condition without interrupting gameplay.

Is `Hit` the best event to attach this to? Would `HandleDamage` or `medical_injured` be different or better in any way?

TODO:

- [ ] Add setting to enable/disable
- [ ] Add setting for duration
- [ ] Fix flickering on subsequent hits before fadeout finishes

- I think this can be fixed by changing the `CBA_fnc_waitAndExecute` to `CBA_fnc_waitUntilAndExecute` and checking the time elapsed since the last hit. Is there an existing, accepted implementation of that? Would there be a problem with multiple identical `CBA_fnc_waitUntilAndExecute` in this situation?

- [ ] Show damaged body part

- Is there a good way to do this? I've searched around a bunch and I haven't been able to figure anything out that would be straightforward. This is more of a hopeful goal at the moment.

**Comment by LinkIsGrim** [comment]
Event should be a matter of preference (as long as execution order is right). Tying to medical's injured event would be most appropriate and consistent with the rest of the components.

**Comment by amsteadrayle** [comment]
This now uses the `medical_woundReceived` event to know about the damaged body part.

Here's a quick demo, tweaked to still show my own info when remote controling a unit.
https://www.youtube.com/watch?v=VGBjroxGD70

File: addons/medical_gui/XEH_postInit.sqf [hunk 1]
```diff
@@ -76,3 +78,20 @@ GVAR(selfInteractionActions) = [];
 ["ace_interactMenuClosed", {
     QGVAR(RscPatientInfo) cutFadeOut 0.3;
 }] call CBA_fnc_addEventHandler;
+
+[QEGVAR(medical,woundReceived), {
+    params ["_unit", "_allDamages", ""];
+    if !(GVAR(peekMedicalOnHit) && ACE_PLAYER == _unit) exitWith {};
```
Comment (rautamiekka) [inline addons/medical_gui/XEH_postInit.sqf:84]:
```suggestion
    if !(GVAR(peekMedicalOnHit) && ACE_player == _unit) exitWith {};
```

File: addons/medical_gui/XEH_postInit.sqf [hunk 2]
```diff
@@ -76,3 +78,20 @@ GVAR(selfInteractionActions) = [];
...
+
+    private _bodypart = toLower (_allDamages select 0 select 1);
```
Comment (rautamiekka) [inline addons/medical_gui/XEH_postInit.sqf:86]: There're only ANSI chars, right ?
```suggestion
    private _bodypart = toLowerANSI (_allDamages select 0 select 1);
```

**Review by rautamiekka — CHANGES_REQUESTED** [review]

File: addons/medical_gui/XEH_postInit.sqf (hunk 2 above)
Comment (amsteadrayle) [reply addons/medical_gui/XEH_postInit.sqf:86]: Probably, but I was just copying the existing usage from here:
https://github.com/acemod/ACE3/blob/78bb0514d1233712428da023cdc732c037ed8aeb/addons/medical_damage/functions/fnc_woundsHandlerBase.sqf#L44

Comment (rautamiekka) [reply addons/medical_gui/XEH_postInit.sqf:86]: Ya, seen exactly that one before and it annoys me a lot.

Comment (LinkIsGrim) [reply addons/medical_gui/XEH_postInit.sqf:86]: Why this? We don't do it anywhere else.

File: addons/medical_gui/XEH_postInit.sqf [hunk 3]
```diff
@@ -76,3 +78,20 @@ GVAR(selfInteractionActions) = [];
 ["ace_interactMenuClosed", {
     QGVAR(RscPatientInfo) cutFadeOut 0.3;
 }] call CBA_fnc_addEventHandler;
+
+[QEGVAR(medical,woundReceived), {
+    params ["_unit", "_allDamages", ""];
+    if !(GVAR(peekMedicalOnHit) && ACE_player == _unit) exitWith {};
```
Comment (LinkIsGrim) [inline addons/medical_gui/XEH_postInit.sqf:84]:
```suggestion
    if !(GVAR(peekMedicalOnHit) && {_unit == ACE_player}) exitWith {};
```

File: addons/medical_gui/initSettings.sqf [hunk 4]
```diff
@@ -126,3 +126,21 @@ private _categoryColors = [ELSTRING(medical,Category), format ["| %1 |", LELSTRI
...
 ] call CBA_fnc_addSetting;
+
+[
+    QGVAR(peekMedicalOnHit),
+    "CHECKBOX",
+    [LSTRING(PeekMedicalOnHit_DisplayName), LSTRING(PeekMedicalOnHit_Description)],
+    [ELSTRING(medical,Category), LSTRING(SubCategory)],
+    true,
+    false // isGlobal
+] call CBA_fnc_addSetting;
```
Comment (LinkIsGrim) [inline addons/medical_gui/initSettings.sqf:137]: This should not be local IMO.

Some might prefer defaulting to off.

Comment (amsteadrayle) [reply addons/medical_gui/initSettings.sqf:137]: I set it local to allow servers to leave it to player preference. Particularly thinking about if somone wants to disable it while a server allows it. But I don't feel too strongly about it.

I do agree with defaulting to off, on second thought.

Comment (veteran29) [reply addons/medical_gui/initSettings.sqf]: Local can always be overwritten.

**Comment by LinkIsGrim** [comment]
Looks good, solve merge conflicts.

File: addons/medical_gui/initSettings.sqf [hunk 5]
```diff
@@ -135,3 +135,21 @@ private _categoryColors = [ELSTRING(medical,Category), format ["| %1 |", LELSTRI
...
+    false,
+    false // isGlobal
+] call CBA_fnc_addSetting;
+
+[
+    QGVAR(peekMedicalOnHitDuration),
+    "TIME",
+    [LSTRING(PeekMedicalOnHitDuration_DisplayName), LSTRING(PeekMedicalOnHitDuration_Description)],
+    [ELSTRING(medical,Category), LSTRING(SubCategory)],
+    [1, 10, 3],
```
Comment (LinkIsGrim) [inline addons/medical_gui/initSettings.sqf:153]:
```suggestion
    [0, 5, 1],
```

**Review by LinkIsGrim — APPROVED** [review]

**Comment by amsteadrayle** [comment]
Those setting limits were intentional. 0 seconds is too short to actually parse any information, and 1 second is still very short. The upper limit being 10 seconds was just an extreme in case anyone felt like looking for that long.
The timing is different from the manual peek because you don't control when or how long it appears.

**Comment by LinkIsGrim** [comment]
There's no harm in reducing the lower end of the time range.

10 seconds is too much time IMO. 5 is an eternity during a firefight already.

I guess default time could've been kept as it was but since this feature is off by default, anyone using it will be looking at settings anyway.

**Pull request merged**